lib/std/fmt.zig

```diff
@@ -1954,7 +1954,519 @@ test "null" {
+    var bytes: [1]u8 = undefined;
+    _ = try reader.read(bytes[0..]);
+    // if(scan_debug) std.debug.print("readByte '{c}'\n", .{ bytes[0] });
+    return bytes[0];
+}
+
+inline fn matchByte(reader: anytype, comptime byte: u8) !bool {
+    const matcher = struct {
+        fn f(c: u8) bool {
+            if (scan_debug) std.debug.print("matcher returnint c '{c}' == byte '{c}' {}\n", .{ c, byte, c == byte });
+            return c == byte;
+        }
+    }.f;
+    return if (try matchByteFn(reader, matcher, true)) |_| true else false;
+}
+
+inline fn matchByteFn(reader: anytype, comptime matchFn: fn (u8) bool, comptime should_match: bool) !?u8 {
+    const byte = try readByte(reader);
+    if (scan_debug) std.debug.print("matchByteFn '{c}' {}\n", .{ byte, byte });
+    if (should_match == matchFn(byte)) return byte;
+    if (scan_debug) std.debug.print("matchByteFn putting back byte '{c}'\n", .{byte});
+    try reader.putBackByte(byte);
+    return null;
+}
+
+inline fn matchBytes(reader: anytype, comptime bytes: []const u8) !bool {
+    inline for (bytes) |byte, i| {
+        if (!try matchByte(reader, byte)) {
+            if (i > 0) {
+                if (scan_debug) std.debug.print("matchBytes putting back bytes '{}'\n", .{bytes[0..i]});
+                try reader.putBack(bytes[0..i]);
+            }
+            return false;
+        }
+    }
+    return true;
+}
+
+// TODO not sure about naming
+pub fn scan(input: []const u8, comptime fmt: []const u8, args: anytype) ScanError!void {
+    if (scan_debug) std.debug.print("scan input '{}' fmt '{}'\n", .{ input, fmt });
+    var fbs = std.io.fixedBufferStream(input);
+    return scanReader(fbs.reader(), fmt, args);
+}
+
+// TODO not sure about naming
+pub fn scanReader(backing_reader: anytype, comptime fmt: []const u8, args: anytype) ScanError!void {
+    if (@typeInfo(@TypeOf(args)) != .Struct) {
+        @compileError("Expected tuple or struct argument, found " ++ @typeName(@TypeOf(args)));
+    }
+    if (args.len > @typeInfo(std.fmt.ArgSetType).Int.bits) {
+        @compileError("32 arguments max are supported per format call");
+    }
+
+    comptime var start_index = 0;
+    comptime var state = State.Start;
+    comptime var maybe_pos_arg: ?comptime_int = null;
+    comptime var specifier_start = 0;
+    comptime var specifier_end = 0;
+    comptime var options = std.fmt.FormatOptions{};
+    comptime var arg_state: std.fmt.ArgState(args.len) = .{};
+
+    var reader = std.io.peekStream(16, backing_reader);
+    var scan_error: ?ScanError = null;
+
+    inline for (fmt) |c, i| {
+        // if(scan_debug) std.debug.print("state {} fmtc '{c}'\n", .{ state, c });
+        if (scan_error != null) return scan_error.?;
+        switch (state) {
+            .Start => switch (c) {
+                '{' => {
+                    if (start_index < i) {
+                        //try writer.writeAll(fmt[start_index..i]);
+                        // TODO not sure what to do here
```

Comment:
@LemonBoy any idea what should be done here?  I don't understand what it means when `start_index < i`.